src/client/mod.rs

```diff
@@ -32,39 +32,120 @@ use status::StatusClass::Redirection;
+    pub can_have_body: bool,
+}
+
+/// A trait that represents the functionality of creating a brand new fresh
+/// request (an instance of a type implementing `FreshHttpRequest`).
+///
+/// Each particular implementation of the trait can provide a different way
+/// for achieving this, as well as a custom type of request that it produces.
+///
+/// Therefore, by defining the type of the fresh request that it will produce,
+/// it also directly defines the type of the response, since the chain of
+/// `Fresh -> Streaming -> Response` types is always well defined given a
+/// particular `FreshHttpRequest` starting point.
+///
+/// TODO: Might need a better name?
+pub trait HttpRequestFactory {
+    /// The type of the `FreshHttpRequest` that this factory will produce.
+    type RequestType: FreshHttpRequest;
+    /// Creates a new instance of the particular `HttpRequestFactory`.
+    /// This method is a convenience so that the `Client` can still be
+    /// instantiated without providing a factory instance, but with a simple
+    /// `new()` call. This, however, means that all particular
+    /// `HttpRequestFactory` implementations must have a set of default
+    /// parameters with which they can be created.
+    fn new() -> Self;
+    /// Creates a new `FreshHttpRequest` of the appropriate type given the
+    /// `RequestTemplate` instance.
+    fn get_fresh_request(&mut self, template: &RequestTemplate) -> HttpResult<Self::RequestType>;
+    /// Set a `ContextVerifier` callback that should be called if the request
+    /// is an HTTPS request.
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier);
+}
+
+/// A struct that represents the functionality of creating new HTTP/1.1
+/// requests.
+pub struct Http11RequestFactory {
+    /// The `Connector` instance used to establish a new connection to
+    /// servers.
+    connector: Connector,
+}
+
+impl Http11RequestFactory {
+    pub fn with_connector(c: Connector) -> Http11RequestFactory {
+        Http11RequestFactory {
+            connector: c,
+        }
+    }
+}
+
+impl HttpRequestFactory for Http11RequestFactory {
+    type RequestType = Request<Fresh>;
+
+    fn new() -> Http11RequestFactory {
+        Http11RequestFactory::with_connector(with_connector(HttpConnector(None)))
+    }
+
+    fn get_fresh_request(&mut self, template: &RequestTemplate) -> HttpResult<Request<Fresh>> {
+        let req = try!(Request::with_connector(
+                template.method.clone(),
+                template.url.clone(),
+                &mut self.connector));
+
+        Ok(req)
+    }
+
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier) {
+        self.connector = with_connector(HttpConnector(Some(verifier)));
+    }
+}
+
 /// A Client to use additional features with Requests.
 ///
 /// Clients can handle things such as: redirect policy.
-pub struct Client {
-    connector: Connector,
+pub struct Client<F=Http11RequestFactory> where F: HttpRequestFactory {
```

Comment:
Hm, you're right, changing that is unfortunate.

------

Here's an idea I haven't fully explored, but what I was thinking about before your pull request arrived in my inbox.

Instead of Client using a RequestFactory, I was thinking that Client would continue to use `client::Request` and `client::Response`. However, those would internally be altered to contain a sort of `HttpMessage`. I imagined it being something like this:

```rust
trait HttpMessage: Read + Write {
    fn get_incoming<T>(&mut self) -> HttpResult<http::MessageHead<T>>;
    fn set_outgoing<T>(&mut self, outgoing: http::MessageHead<T>) -> HttpResult<()>;
}
```

Request and Response would be altered:

```rust
struct Request<W> {
    message: Box<HttpMessage>,
    // ...
}

impl Request<Fresh> {
    pub fn start(self) -> HttpResult<Request<Streaming>> {
        try!(self.message.set_outgoing(MessageHead {
            subject: (self.method, self.uri),
            headers: self.headers,
            // version seems redundant?
        }));
    }
}

// and so forth...
```

The `HttpMessage` implementation would know how to set up its internal Read and Write. For H1, the implementation would adjust itself upon receiving an `Outgoing` to determine which `HttpWriter` to use.

There's currently a `http::Incoming`, that could be adjusted to work for both incoming and outgoing. Possible name is `MessageHead`.